lib/libuzfs/libuzfs.c

```diff
@@ -1392,45 +1393,61 @@ int
 libuzfs_object_write(libuzfs_dataset_handle_t *dhp, uint64_t obj,
     uint64_t offset, uint64_t size, const char *buf, boolean_t sync)
 {
-	int err = 0;
-	objset_t *os = dhp->os;
-	zilog_t *zilog = dhp->zilog;
-	dmu_tx_t *tx = NULL;
-	uzfs_object_attr_t *attr = NULL;
-	dmu_buf_t *db;
+	dnode_t *dn;
+	int err;
+	if ((err = dnode_hold(dhp->os, obj, FTAG, &dn)) != 0) {
+		return (err);
+	}
 
-	err = dmu_bonus_hold(os, obj, FTAG, &db);
-	if (err)
+	dmu_buf_t *bounus_buf;
+	err = dmu_bonus_hold_by_dnode(dn, FTAG,
+	    &bounus_buf, DMU_READ_NO_PREFETCH);
+	if (err != 0) {
+		dnode_rele(dn, FTAG);
 		return (err);
+	}
 
-	attr = db->db_data;
+	uzfs_object_attr_t *attr = dn->dn_bonus->db.db_data;
+	uint64_t end_size = offset + size;
+	uint64_t max_blksz = dhp->max_blksz;
+	objset_t *os = dhp->os;
+	while (size > 0) {
+		uint64_t nwrite = MIN(size,
+		    max_blksz - P2PHASE(offset, max_blksz));
+		dmu_tx_t *tx = dmu_tx_create(os);
+		dmu_tx_hold_write_by_dnode(tx, dn, offset, nwrite);
+		if ((err = dmu_tx_assign(tx, TXG_WAIT)) != 0) {
+			dmu_tx_abort(tx);
+			goto out;
+		}
 
-	tx = dmu_tx_create(os);
+		// update max blocksize when current blksz too small
+		if (dn->dn_datablksz < max_blksz &&
+		    dn->dn_datablksz < end_size) {
```

Comment:
why is block size (a relative value) compared with end_size (an absolute value)?